parsing/pprintast.ml

```diff
@@ -81,13 +81,23 @@ let last_is c str =
 let first_is_in cs str =
   str <> "" && List.mem str.[0] cs
 
+(* Classify intersection with lowercase identifiers and operators due to raw
+   identifiers *)
+type longindent_kind =
+  | Constr (* [true] and [false] *)
+  | Type  (* [mod] *)
+  | Other
```

Comment:
I feel that I understand almost everything in this PR, except for this type declaration and its documentation.
- I don't know what the comment ("Classify ...") means.
- The documentation is specifically talking about raw identifiers, but I think that the meaning of this "kind" is independent from raw identifiers. (It's some information that we need to track because of raw identifier printing needs, but which is independent from it.)
- My reverse-engineering guess is that this is pretty-printing *context* information, that carries information on the grammatical context in which the printed value will be used. Another way to view it is that we are specifying the grammatical category of what is being printed. (For a notion of grammar that is finer-grained than just "longident", obviously.) It seems to be related to (a coarser-grained version of) the type `Out_type.namespace`. 

I think that if the documentation of this type (and maybe the type name?) was changed to better reflect my reverse-engineered guess, then the PR would look very nice to me.

Reply:
Yes, this is too distilled comments on the fact that `Longident` is a too coarse grammatical category to allow unparsing: the grammar rules for constructor, module, type and value longidents are different. Before the introduction of raw identifiers, we could recover just enough information to print them correctly by recognizing special case in each category (`true`  and `false` for constructors, operators for values).

Raw identifiers makes this context-less reconstruction impossible by making it possible for those special case to appear escaped in other kind of longidents. For instance `mod` might be either an operator name in a value longident, or a raw identifier in a type longident. Similarly, `true` might be a normal constructor or a raw identifier in a value longident.

Thus this PR uses contextual information to recover the information about which form of longidents we are printing and how should we print them according to their specific grammar.

I will amend the comment to be more explicit (next week).